### DISCUSSION TITLE
Support default values for "executemany" insertion of dictionaries with sparsely defined column values

### DISCUSSION THREAD

**Discussion by chrispy-snps** [sqlalchemy/sqlalchemy#11908] [Ideas]
Currently, when "executemany" insertion of a Python list of dictionaries is performed, the relevant parameters are taken from the first dictionary in the list.

Consider the following table definition:

```python
# Define a table with Python-side default values
engine = sqlalchemy.create_engine("...")
metadata = sqlalchemy.MetaData()
table1 = sqlalchemy.Table(
    "table1",
    metadata,
    sqlalchemy.Column("a", sqlalchemy.Integer, default=9),
    sqlalchemy.Column("b", sqlalchemy.Integer, default=9),
    sqlalchemy.Column("c", sqlalchemy.Integer, default=9),
)
metadata.create_all(engine)
```

If I push the following data:

```python
with engine.begin() as connection:
    connection.execute(
        sqlalchemy.insert(table1),
        [
            {"a": 1, "b": 2},
            {"a": 3, "b": 4, "c": 5},
        ],
    )
```

then the insertion considers only columns "a" and "b" per the first row, and "c" assumes the default of 9 in the second row despite the explicitly provided value:

| a | b | c |
| - | - | - |
| 1 | 2 | 9 |
| 3 | 4 | **9** |

If I push the following data:

```python
with engine.begin() as connection:
    connection.execute(
        sqlalchemy.insert(table1),
        [
            {"a": 1, "b": 2},
            {"a": 3, "c": 5},
        ],
    )
```

then the insertion considers only columns "a" and "b" per the first row, and the transaction fails because "b" is undefined in the second row:

```
StatementError: (sqlalchemy.exc.InvalidRequestError) A value is required for bind parameter 'b', in parameter group 1
[SQL: INSERT INTO table1 (a, b, c) VALUES (%(a)s, %(b)s, %(c)s)]
[parameters: [{'a': 1, 'b': 2}, {'a': 3, 'c': 5}]]
(Background on this error at: https://sqlalche.me/e/20/cd3x)
```

(The writeup for `cd3x` is quite clear, by the way.)

It would be a nice enhancement for SQLAlchemy to scan all dictionary rows to determine what parameters are involved, maybe something like this:

```python
all_parameters = set().union(*(d.keys() for d in data))
```

then use the Python-side defaults for dictionary items that are not provided (or issue `cd3x` with modified guidance that another possible solution is to provide `default` values for possibly-missing fields).

(Although perhaps if it was that simple, it would already be done by now!)

**Comment by CaselIT** [comment]
Hi,

This is not done for performance reason. Processing each dictionary one by one will become quite a bit slow, especially if the list side is in the thousands

**Comment by chrispy-snps** [comment]
@CaselIT - thanks for the reply!

Given 100,000 rows of 20 randomly-named fields each:

```python
import random

data = []
for i in range(100000):
    data.append(
        {
            f"f{random.randint(1, 100)}": True for _ in range(20)
        }
    )
```

the runtime to determine the set of referenced keys is 50ms on my machine:

```python
import time
start = time.time()
all_parameters = set().union(*(d.keys() for d in data))
print(f"Elapsed time: {time.time() - start}")
```

This does not seem too bad relative to the CPU/network/disk processing required for the rest of the transaction, although it is admittedly an overhead that would be applied everywhere unless enabled by a parameter.

**Reply by CaselIT** [reply]
I would think that if such a feature were available people would expect different statement for the different keys. It gets complicated quickly

**Reply by chrispy-snps** [reply]
@CaselIT - I appreciate the consideration. I am new to both SQL and SQLAlchemy, so it's hard for me to see the potential complexity in my request.

For now, this approach gets the job done:

```python
column_defaults = {
    "a": 9,
    "b": 9,
    "c": 9,
}

with self.engine.begin() as conn:
    conn.execute(
        data_table.insert().values(
            [
                {f: (row[f] if f in row else d) for f, d in column_defaults.items()}
                for row in data
            ]
        )
    )
```

**Comment by zzzeek** [comment]
this does get complicated quickly and it's mostly because we have three varieties of "defaults", only one of which actually goes in the parameter dictionary, and additionally that scanning all the parameter sets for every bulk insert before we can decide to compile the statement would be a mostly unwelcome and highly variable / surprising performance hit for all applications globally.

So due to the second item mostly, in this case, the thing you want is something you absolutely can simply do yourself ahead of time, for those statements where you aren't sure what kind of keys you want in your dictionaries.    Several examples here, here's the simple version

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import engine
from sqlalchemy import insert
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import Table


# Define a table with Python-side default values
engine = create_engine("sqlite://", echo=True)

metadata = MetaData()
table1 = Table(
    "table1",
    metadata,
    Column("a", Integer),
    Column("b", Integer, default=9),
    Column("c", Integer, default=9),
)
metadata.create_all(engine)


def normalize_params(table, data):
    all_parameters = set().union(*(d.keys() for d in data))
    client_side_defaults = {
        k: c.default.arg
        for k, c in (((c.key, c) for c in table.c))
        if c.default is not None
        and c.default.is_scalar  # can be expanded to support python side
        # callables, but only if they dont accept "context"
    }
    return [
        {
            k: d[k] if k in d else client_side_defaults[k]
            for k in all_parameters
        }
        for d in data
    ]


with engine.begin() as connection:
    connection.execute(
        insert(table1),
        normalize_params(
            table1,
            [
                {"a": 1, "b": 2},
                {"a": 3, "b": 4, "c": 5},
            ],
        ),
    )

```

turns out we can set normalize_params into our own execution option so that you can opt in to this additional complexity if you want it, using an event:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import engine
from sqlalchemy import event
from sqlalchemy import insert
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import Table


# Define a table with Python-side default values
engine = create_engine("sqlite://", echo=True)

metadata = MetaData()
table1 = Table(
    "table1",
    metadata,
    Column("a", Integer),
    Column("b", Integer, default=9),
    Column("c", Integer, default=9),
)
metadata.create_all(engine)


@event.listens_for(engine, "before_execute")
def receive_before_execute(
    conn, clauseelement, multiparams, params, execution_options
):
    if "normalize_insert_params" not in execution_options:
        return

    table = clauseelement.table
    all_parameters = set().union(*(d.keys() for d in multiparams))
    client_side_defaults = {
        k: c.default.arg
        for k, c in (((c.key, c) for c in table.c))
        if c.default is not None
        and c.default.is_scalar  # can be expanded to support python side
        # callables, but only if they dont accept "context"
    }
    multiparams[:] = [
        {
            k: d[k] if k in d else client_side_defaults[k]
            for k in all_parameters
        }
        for d in multiparams
    ]


with engine.begin() as connection:
    connection.execute(
        insert(table1),
        [
            {"a": 1, "b": 2},
            {"a": 3, "b": 4, "c": 5},
        ],
        execution_options=dict(normalize_insert_params=True)
    )

```

then we have the looks of something that could be a wiki recipe, a third party plugin or if eventually we had extremely comprehensive tests for a very wide range of cases with clear failure modes for uncorrectable cases, a SQLAlchemy extension

**Reply by chrispy-snps** [reply]
@zzzeek - this is great! I appreciate you walking through the use of the `Column` default values first, then extending that to event-based application. I learned new things in both areas.

Now there is plenty of information here for others who need this functionality. And as you say, these solutions could be more widely distributed over time if needed.

**Reply by zzzeek** [reply]
yes, note the event handler makes a ton of assumptions, namely that the "clauseelement" is an "insert" that has a ".table" on it.  in "the real world", this execution option might be there for other kinds of statements as well and it would then need to be ignored / do something for UPDATES ? / warn / raise / etc, who knows.
